File: backend/src/cli/commands.ts
import type { StartServerResult } from '@/app/server'
import { openIpSelector } from '@/app/server'
import { internalConfig } from '@/config/config'
import { opener } from '@/utils/server-utils.ts'

export interface CmdCtx {
  serverResult?: StartServerResult
  createConfigAndReload: () => Promise<void>
  reload: () => Promise<void>
  exit: () => Promise<void>
}

export interface Command {
  id: string
  label: string
  when?: (ctx: CmdCtx) => boolean
  run: (ctx: CmdCtx) => Promise<void>
}

function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

export const commands: Command[] = [
  {
    id: 'open-ip',
    label: 'Open IP selector',
    when: ctx => Boolean(ctx.serverResult),
    async run(ctx) {
      if (ctx.serverResult) {
        await openIpSelector(ctx.serverResult)
      }
      await sleep(1000)
    },
  },
  {
    id: 'print-urls',
    label: 'Print urls',
    when: ctx => Boolean(ctx.serverResult),
    async run(ctx) {
      console.clear()
      ctx.serverResult?.printUrls()
    },
  },
  {
    id: 'create-config',
    label: 'Create config file',
    when: () => !internalConfig.configInitialized,
    async run(ctx) {
      await ctx.createConfigAndReload()
    },
  },
  {
    id: 'open-config',
    label: 'Open config file',
    when: () => internalConfig.configInitialized,
    async run() {
      if (internalConfig.configInitialized) {
        await opener(internalConfig.configFilePath)
      }
    },
  },
  {
    id: 'reload',
    label: 'Restart server',
    when: ctx => Boolean(ctx.serverResult),
    async run(ctx) {
      console.clear()
      await ctx.reload()
    },
  },
  {
    id: 'exit',
    label: 'Exit',
    async run(ctx) {
      await ctx.exit()
    },
  },
]

File: backend-go/cli/commands.go
package cli

import (
	"fmt"
	"time"

	"file-lite-go/config"
	"file-lite-go/utils"
)

type ServerResult struct {
	UrlIpSelector string
	PrintUrls     func()
}

type CmdCtx struct {
	ServerResult          *ServerResult
	CreateConfigAndReload func() error
	Reload                func() error
	Exit                  func() error
}

type Command struct {
	ID    string
	Label string
	When  func(ctx CmdCtx) bool
	Run   func(ctx CmdCtx) error
}

func Commands() []Command {
	return []Command{
		{
			ID:    "open-ip",
			Label: "Open IP selector",
			When:  func(ctx CmdCtx) bool { return ctx.ServerResult != nil },
			Run: func(ctx CmdCtx) error {
				if ctx.ServerResult != nil {
					ctx.ServerResult.PrintUrls()
					utils.Opener(ctx.ServerResult.UrlIpSelector)
				}
				time.Sleep(time.Second)
				return nil
			},
		},
		{
			ID:    "print-urls",
			Label: "Print urls",
			When:  func(ctx CmdCtx) bool { return ctx.ServerResult != nil },
			Run: func(ctx CmdCtx) error {
				fmt.Print("\033[H\033[2J")
				if ctx.ServerResult != nil {
					ctx.ServerResult.PrintUrls()
				}
				return nil
			},
		},
		{
			ID:    "create-config",
			Label: "Create config file",
			When:  func(CmdCtx) bool { return !config.ConfigInitialized() },
			Run: func(ctx CmdCtx) error {
				return ctx.CreateConfigAndReload()
			},
		},
		{
			ID:    "open-config",
			Label: "Open config file",
			When:  func(CmdCtx) bool { return config.ConfigInitialized() },
			Run: func(CmdCtx) error {
				if config.ConfigInitialized() {
					utils.Opener(config.ConfigFilePath())
				}
				return nil
			},
		},
		{
			ID:    "reload",
			Label: "Restart server",
			When:  func(ctx CmdCtx) bool { return ctx.ServerResult != nil },
			Run: func(ctx CmdCtx) error {
				fmt.Print("\033[H\033[2J")
				return ctx.Reload()
			},
		},
		{
			ID:    "exit",
			Label: "Exit",
			Run: func(ctx CmdCtx) error {
				return ctx.Exit()
			},
		},
	}
}
